Update and fix `ledger-state` benchmarks

neilmayhew opened the pull request:
# Description

* The `performance` benchmark in `ledger-state` has been broken since the start of Conway
* The instructions in the `README` are out of date
* Help and progress information is inaccurate
* We can now take advantage of `mithril` to speed up the process of getting a ledger snapshot
* Snapshots are different since the introduction of UTxO-HD in node `10.4`

These changes include a temporary workaround for the fact that the pre-stored transactions are no longer valid in Conway. With this workaround it's possible to run the benchmarks but they're measuring failing transactions which isn't representative of real traffic. When the transactions are fixed the workaround can be removed (eg with `git revert` since they're in their own commit).

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [ ] Tests added or updated when needed.
- [ ] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [ ] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

neilmayhew committed:
Fix and update ledger-state benchmark instructions

neilmayhew committed:
Don't crash when UTxO is too small

neilmayhew committed:
Work around decoding issue in hardcoded transactions

The decoding error is:

> TxBody: 'Collateral Inputs' must be non-empty when supplied

neilmayhew committed:
Handle need for separate UTxO query with cardano-node 10.4

neilmayhew committed:
Add a multi-tx applyTx benchmark

neilmayhew committed:
Allow predicate failures from applyTx

This is a temporary measure until the embedded transactions are fixed
for Conway. With this change it's possible to run the benchmarks but
they're measuring failing transactions which isn't representative of
real traffic.

neilmayhew committed:
Add a script for downloading mithril snapshots

Also mention using mithril in the ledger-state README

lehins commented on libs/ledger-state/bench/Performance.hs:
```diff
@@ -176,10 +199,12 @@ selectRandomMapKeys n gen m = runStateGenT_ gen $ \g ->
 decodeTx hex = either error id $ do
   bsl <- BSL16.decode hex
-  first show $ decodeFull (eraProtVerHigh @CurrentEra) bsl
+  tx <- first show $ decodeFull (eraProtVerHigh @BabbageEra) bsl
+  -- TODO: remove this after the transactions below are updated
+  first show $ upgradeTx tx
```
This doesn't make sense, since transaction should still be deserializable in the ConwayEra. Did you get a decoder error when you used `CurrentEra`?

There is a chance that you  can get a predicate failure when you apply the Tx, since I might have accidentally spent the inputs that are used in those test transactions :see_no_evil: In which case we'd need to recreate those transactions.

neilmayhew replied:
This the decoder error in Conway:

> TxBody: 'Collateral Inputs' must be non-empty when supplied

It's generated [here](https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/conway/impl/src/Cardano/Ledger/Conway/TxBody.hs#L194).

By decoding in Babbage and then upgrading we avoid the decoding failure but then get a predicate failure because the witness signature is no longer valid. However, this enables us to run the affected benchmarks by accepting an `Either` result whereas the decoding failure is unrecoverable because there's no tx to apply.

I wasn't sure whose `TxOut`s were being spent, but I guessed they were yours :grinning: They're still unspent!

lehins commented on libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs:
```diff
@@ -57,11 +61,20 @@ readEpochState ::
+readUTxO ::
+  FilePath ->
+  IO (UTxO CurrentEra)
+readUTxO = readDecCBORHex
```
Why do you expect UTxO to be hex encoded?
Even the scripts you created expect a `utxo.bin` file.

That being said, if `cardano-cli query utxo` produces a hex encoded file it would be better to figure out how it can dump cbor directly. And if it can't then changing to `utxo.hex` for the file name would provide more clarity on the contents. Also if hex is the only option then, changing the name to something like `readHexUTxO` would also provide more clarity

neilmayhew replied:
`cardano-cli` does in fact output a hex-encoded file from `query utxo --output-cbor`. (I just had a PR accepted that changes it to use a lazy ByteString instead of a strict one. Quite important when dumping the 3.5GB of the whole UTxO!) I haven't found out why it's using hex rather than binary, but my guess is that the typical use-case is for a partial UTxO filtered using the `--address` or `--tx-in` options and that it's somehow more useful to have hex in those cases.

You're right, I should have used a `.hex` suffix. I'll make that change, and rename the function. I also forgot to add the `--output-cbor` argument in the instructions, so I'll change that too.

We could ask the cli team whether they'd be open to adding another option to that query to specify binary output (eg `--output-cbor-binary`) but I assume not many people are interested in dumping the whole UTxO so it wouldn't be worthwhile.

lehins reviewed (approved):
Thank you for giving some care to the neglected `ledger-state` :smile:

neilmayhew commented:
I updated the the function name, the file name and the cli arguments, but I also added a check to allow only the signature predicate failure and nothing else.
